objectserver/ecengine.go

```diff
@@ -465,43 +534,95 @@ func (f *ecEngine) RegisterHandlers(addRoute func(method, path string, handler h
 	addRoute("PUT", "/ec-reconstruct/:device/:account/:container/*obj", f.ecReconstructHandler)
 }
 
-func (f *ecEngine) GetObjectsToStabilize(device string, c chan ObjectStabilizer, cancel chan struct{}) {
-	defer close(c)
-	idb, err := f.getDB(device)
-	if err != nil {
-		return
+func (f *ecEngine) updateItemsBeingStabilized(device string, objs []*ecObject) {
+	f.stabm.Lock()
+	defer f.stabm.Unlock()
+	delete(f.stabItems, device)
+	f.stabItems[device] = map[string]bool{}
+	for _, o := range objs {
+		f.stabItems[device][fmt.Sprintf("%s-%s", o.Hash, o.metadata["X-Timestamp"])] = true
 	}
+}
 
-	idb.ExpireObjects()
+func (f *ecEngine) UpdateItemStabilized(device, hash, ts string) {
+	f.stabm.Lock()
+	defer f.stabm.Unlock()
+	if _, ok := f.stabItems[device]; ok {
+		f.stabItems[device][fmt.Sprintf("%s-%s", hash, ts)] = false
+	}
+}
 
-	items, err := idb.ListObjectsToStabilize()
+func (f *ecEngine) hasBeenStabilized(device, hash, ts string) bool {
+	f.stabm.Lock()
+	defer f.stabm.Unlock()
+	if _, ok := f.stabItems[device]; ok {
+		if val, ok := f.stabItems[device][fmt.Sprintf("%s-%s", hash, ts)]; !val && ok {
+			return true
+		}
+	}
+	return false
+}
+
+func (f *ecEngine) GetObjectsToStabilize(device *ring.Device, c chan ObjectStabilizer, cancel chan struct{}) {
+	defer close(c)
+	idb, err := f.getDB(device.Device)
 	if err != nil {
 		return
 	}
+	idb.ExpireObjects()
 
-	for _, item := range items {
-		obj := &ecObject{
-			IndexDBItem:     *item,
-			idb:             idb,
-			policy:          f.policy,
-			metadata:        map[string]string{},
-			ring:            f.ring,
-			logger:          f.logger,
-			reserve:         f.reserve,
-			dataShards:      f.dataShards,
-			parityShards:    f.parityShards,
-			chunkSize:       f.chunkSize,
-			client:          f.client,
-			nurseryReplicas: f.nurseryReplicas,
-			txnId:           fmt.Sprintf("%s-%s", common.UUID(), device),
+	idbItemsChan := make(chan *IndexDBItem, numStabilizeObjects)
+	cancel = make(chan struct{})
+	defer close(cancel)
+	go idb.ListObjectsToStabilize(idbItemsChan, cancel)
+	allDone := false
+	for !allDone {
+		objs := []*ecObject{}
+		for i := 0; i < numStabilizeObjects; i++ {
```

Comment:
It's kind of strange using this here AND as a db query limit. Might not be a huge deal, but if db1 has 1000 things, you'll only do those and stop? Maybe that's exactly what's intended...